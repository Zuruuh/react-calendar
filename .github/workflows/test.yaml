name: Tests and static analysis

on:
  push:
    branches: ["dev"]
  pull_request:
    branches: ["dev"]

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: ðŸ›’ Checkout branch
        uses: actions/checkout@v4

      - name: âœ¨ Install bun
        uses: oven-sh/setup-bun@v1
        with:
          bun-version: 1.0.25

      - uses: actions/cache@v4
        name: ðŸ“¦ Setup deps cache
        with:
          path: node_modules
          key: ${{ runner.os }}-modules-${{ hashFiles('**/bun.lockb') }}
          restore-keys: |
            ${{ runner.os }}-modules-

      - name: âž• Install dependencies
        run: bun install --frozen-lockfile

      - name: ðŸš¨ Check formatting
        run: bun run format

      - name: ðŸš¨ Lint
        if: always()
        run: bun run lint

      - name: ðŸ§ª Tests
        if: always()
        run: bun test
