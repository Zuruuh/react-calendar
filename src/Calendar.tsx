import { useCallback, useState, type FC, type ReactNode } from 'react';
import {
  CalendarContext,
  type CalendarOverlap,
  type CalendarState,
} from './context/CalendarContext';
import type { Setter } from './types/Setter';
import { Weeks } from './components/Weeks';
import { Week } from './components/Week';
import { Day } from './components/Day';
import day, { type Dayjs } from 'dayjs';
import { CalendarPlugin } from './plugin';

export interface CalendarProps {
  // selectedDate: Dayjs | null;
  // setSelectedDate: Setter<Dayjs | null>;
  viewedDate?: Dayjs;
  dayjs?(): Dayjs;
  plugins?: Array<CalendarPlugin>;
  // overlap?: CalendarOverlap;
  // altDateFormat?: string;
  children: ReactNode | ((props: CalendarState) => ReactNode);
}

const Calendar: FC<CalendarProps> = (props) => {
  const {
    children,
    // overlap = 'overlap',
    // altDateFormat = 'dddd D MMMM YYYY',
    plugins = [],
    viewedDate,
    dayjs = () => day(),
  } = props;

  const dayFactory = useCallback(
    () => dayjs().utc(true).second(0).minute(0).hour(12),
    [dayjs],
  );

  // const additionnalProps = plugins.map((plugin) => plugin?.calendarHook !== undefined ? plugin.calendarHook() : {});

  // const [temporarySelectedDate, setTemporarySelectedDate] = useState(
  //   dayFactory().day(dayjs().localeData().firstDayOfWeek()),
  // );

  // const setTemporarySelectedDateDecorator: Setter<Dayjs> = (date) => {
  //   const unwrappedDate =
  //     typeof date === 'function' ? date(temporarySelectedDate) : date;
  //   return setTemporarySelectedDate(unwrappedDate.date(1));
  // };

  const state: CalendarState = {
    // temporarySelectedDate,
    // setTemporarySelectedDate: setTemporarySelectedDateDecorator,
    // controls: {
    //   nextMonth: controlFactory(true, 'month', maximumSelectableDate),
    //   nextYear: controlFactory(true, 'year', maximumSelectableDate),
    //   prevMonth: controlFactory(false, 'month', minimumSelectableDate),
    //   prevYear: controlFactory(false, 'year', minimumSelectableDate),
    // },
    // overlap,
    // altDateFormat,
    plugins,
    viewedDate: viewedDate ?? dayFactory(),
    dayjs: dayFactory,
  };

  return (
    <CalendarContext.Provider value={state}>
      {typeof children === 'function' ? children(state) : children}
    </CalendarContext.Provider>
  );
};

export default {
  Root: Calendar,
  Weeks,
  Week,
  Day,
};
