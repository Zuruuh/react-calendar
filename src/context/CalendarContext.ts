import { createContext, useContext } from 'react';
import type { Setter } from '../types/Setter';
import type { Dayjs } from 'dayjs';

export type CalendarOverlap =
  | 'overlap'
  | 'no-overlap'
  | 'no-overlap-with-offset';

export interface CalendarControl {
  execute(): void;
  disabled: boolean;
}

export interface CalendarControls {
  nextMonth: CalendarControl;
  nextYear: CalendarControl;
  prevMonth: CalendarControl;
  prevYear: CalendarControl;
}

export interface CalendarState {
  temporarySelectedDate: Dayjs;
  setTemporarySelectedDate: Setter<Dayjs>;
  minimumSelectableDate: Dayjs;
  maximumSelectableDate: Dayjs;
  controls: CalendarControls;
  overlap: CalendarOverlap;
  dayjs(): Dayjs;
  altDateFormat: string;
}

/**
 * @internal
 */
export const CalendarContext = createContext<CalendarState | undefined>(
  undefined,
);

/**
 * @internal
 */
export function useCalendarContext(): CalendarState {
  const state = useContext(CalendarContext);
  if (state === undefined) {
    throw new Error(
      'Uninitialized date picker context used! ' +
        'You probably tried to render a Date Picker (like <Calendar.Weeks>) ' +
        'element without wrapping it in a <Calendar.Root>',
    );
  }

  return state;
}
