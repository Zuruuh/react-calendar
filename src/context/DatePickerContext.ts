import { createContext, useContext } from 'react';
import type { Setter } from '../types/Setter';
import type { Dayjs } from 'dayjs';

export type DatePickerCalendarOverlap =
  | 'overlap'
  | 'no-overlap'
  | 'no-overlap-with-offset';

export interface DatePickerControl {
  execute(): void;
  disabled: boolean;
}

export interface DatePickerControls {
  nextMonth: DatePickerControl;
  nextYear: DatePickerControl;
  prevMonth: DatePickerControl;
  prevYear: DatePickerControl;
}

export interface DatePickerState {
  selectedDate: Dayjs | null;
  setSelectedDate(date: Dayjs | null): void;
  temporarySelectedDate: Dayjs;
  setTemporarySelectedDate: Setter<Dayjs>;
  minimumSelectableDate: Dayjs;
  maximumSelectableDate: Dayjs;
  controls: DatePickerControls;
  overlap: DatePickerCalendarOverlap;
  dayjs(): Dayjs;
}

/**
 * @internal
 */
export const DatePickerContext = createContext<DatePickerState | undefined>(
  undefined,
);

/**
 * @internal
 */
export function useDatePickerContext(): DatePickerState {
  const state = useContext(DatePickerContext);
  if (state === undefined) {
    throw new Error(
      'Uninitialized date picker context used! ' +
        'You probably tried to render a Date Picker element without wrapping it in a <DatePicker.Root>',
    );
  }

  return state;
}
