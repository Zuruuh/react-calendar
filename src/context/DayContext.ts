import type { Dayjs } from 'dayjs';
import { createContext, useContext } from 'react';

export interface DayContextState {
  date: Dayjs;
}

/**
 * @internal
 */
export const DayContext = createContext<DayContextState | undefined>(undefined);

/**
 * @internal
 */
export function useDayContext(): DayContextState {
  const state = useContext(DayContext);
  if (state === undefined) {
    throw new Error(
      'Uninitialized Day context used! ' +
        'You probably tried to render a <DatePicker.Day> element without wrapping it in a <DatePicker.Week>'
    );
  }

  return state;
}
