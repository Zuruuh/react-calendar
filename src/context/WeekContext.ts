import { createContext, useContext } from 'react';
import type { WeekNumber } from '../types/WeekNumber';

export interface WeekContextState {
  weekNumbers: WeekNumber;
}

/**
 * @internal
 */
export const WeekContext = createContext<WeekContextState | undefined>(
  undefined
);

/**
 * @internal
 */
export function useWeekContext(): WeekContextState {
  const state = useContext(WeekContext);
  if (state === undefined) {
    throw new Error(
      'Uninitialized week context used! ' +
        'You probably tried to render a <DatePicker.Week> element without wrapping it in a <DatePicker.Calendar>'
    );
  }

  return state;
}
