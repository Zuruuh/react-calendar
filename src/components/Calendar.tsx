import { useCallback } from 'react';
import type { ReactNode, FC } from 'react';
import { useDatePickerContext } from '../context/DatePickerContext';
import { WeekContext } from '../context/WeekContext';

export interface CalendarInnerProps {
  weekNumber: number;
}

export interface CalendarProps {
  children: ReactNode | ((props: CalendarInnerProps) => ReactNode);
}

export const Calendar: FC<CalendarProps> = ({ children }) => {
  const { temporarySelectedDate } = useDatePickerContext();
  const referenceDate = temporarySelectedDate;
  const startOfMonth = referenceDate.date(1).endOf('week');

  const createChildren = useCallback(
    (props: CalendarInnerProps) =>
      typeof children === 'function' ? children(props) : children,
    [children]
  );

  const weeks: CalendarInnerProps[] = [{ weekNumber: startOfMonth.week() }];

  for (
    let i = 0;
    startOfMonth.add(i, 'week').month() === referenceDate.month();
    i++
  ) {
    const weekDate = startOfMonth.add(i + 1, 'week').startOf('week');
    const weekNumber = weekDate.week();

    if (weekDate.month() === referenceDate.month()) {
      weeks.push({ weekNumber });
    }
  }

  return (
    <>
      {weeks.map(({ weekNumber }) => (
        <WeekContext.Provider key={weekNumber} value={{ weekNumber }}>
          {createChildren({ weekNumber })}
        </WeekContext.Provider>
      ))}
    </>
  );
};
