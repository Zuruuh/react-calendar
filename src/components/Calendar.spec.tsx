import { test, expect } from 'bun:test';
import { generateWeeksBasedOnOverlap } from './Calendar';
import { type Dayjs } from 'dayjs';
import { type DatePickerCalendarOverlap } from '../context/DatePickerContext';
import { day } from '../tests/utils';

type Output = [Dayjs, DatePickerCalendarOverlap];

const expectedOutputs: Output[] = [
  [day({ year: 2023, month: 1, day: 1 }), 'overlap'],
  [day({ year: 2023, month: 2, day: 1 }), 'overlap'],
  [day({ year: 2023, month: 3, day: 1 }), 'overlap'],
  [day({ year: 2023, month: 4, day: 1 }), 'overlap'],
  [day({ year: 2023, month: 5, day: 1 }), 'overlap'],
  [day({ year: 2023, month: 6, day: 1 }), 'overlap'],
  [day({ year: 2023, month: 7, day: 1 }), 'overlap'],
  [day({ year: 2023, month: 8, day: 1 }), 'overlap'],
  [day({ year: 2023, month: 9, day: 1 }), 'overlap'],
  [day({ year: 2023, month: 10, day: 1 }), 'overlap'],
  [day({ year: 2023, month: 11, day: 1 }), 'overlap'],
  [day({ year: 2023, month: 12, day: 1 }), 'overlap'],

  [day({ year: 2023, month: 1, day: 1 }), 'no-overlap'],
  [day({ year: 2023, month: 2, day: 1 }), 'no-overlap'],
  [day({ year: 2023, month: 3, day: 1 }), 'no-overlap'],
  [day({ year: 2023, month: 4, day: 1 }), 'no-overlap'],
  [day({ year: 2023, month: 5, day: 1 }), 'no-overlap'],
  [day({ year: 2023, month: 6, day: 1 }), 'no-overlap'],
  [day({ year: 2023, month: 7, day: 1 }), 'no-overlap'],
  [day({ year: 2023, month: 8, day: 1 }), 'no-overlap'],
  [day({ year: 2023, month: 9, day: 1 }), 'no-overlap'],
  [day({ year: 2023, month: 10, day: 1 }), 'no-overlap'],
  [day({ year: 2023, month: 11, day: 1 }), 'no-overlap'],
  [day({ year: 2023, month: 12, day: 1 }), 'no-overlap'],

  [day({ year: 2023, month: 1, day: 1 }), 'no-overlap-with-offset'],
  [day({ year: 2023, month: 2, day: 1 }), 'no-overlap-with-offset'],
  [day({ year: 2023, month: 3, day: 1 }), 'no-overlap-with-offset'],
  [day({ year: 2023, month: 4, day: 1 }), 'no-overlap-with-offset'],
  [day({ year: 2023, month: 5, day: 1 }), 'no-overlap-with-offset'],
  [day({ year: 2023, month: 6, day: 1 }), 'no-overlap-with-offset'],
  [day({ year: 2023, month: 7, day: 1 }), 'no-overlap-with-offset'],
  [day({ year: 2023, month: 8, day: 1 }), 'no-overlap-with-offset'],
  [day({ year: 2023, month: 9, day: 1 }), 'no-overlap-with-offset'],
  [day({ year: 2023, month: 10, day: 1 }), 'no-overlap-with-offset'],
  [day({ year: 2023, month: 11, day: 1 }), 'no-overlap-with-offset'],
  [day({ year: 2023, month: 12, day: 1 }), 'no-overlap-with-offset'],
];

test('generateWeeksBasedOnOverlap', () => {
  expectedOutputs.forEach(([date, type]) => {
    expect(
      Array.from(generateWeeksBasedOnOverlap(date, type)),
    ).toMatchSnapshot();
  });
});
