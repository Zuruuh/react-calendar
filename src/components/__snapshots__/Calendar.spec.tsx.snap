// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`generateWeeksBasedOnOverlap 1`] = `
[
  [
    52,
  ],
  [
    1,
  ],
  [
    2,
  ],
  [
    3,
  ],
  [
    4,
  ],
  [
    5,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 2`] = `
[
  [
    5,
  ],
  [
    6,
  ],
  [
    7,
  ],
  [
    8,
  ],
  [
    9,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 3`] = `
[
  [
    9,
  ],
  [
    10,
  ],
  [
    11,
  ],
  [
    12,
  ],
  [
    13,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 4`] = `
[
  [
    13,
  ],
  [
    14,
  ],
  [
    15,
  ],
  [
    16,
  ],
  [
    17,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 5`] = `
[
  [
    18,
  ],
  [
    19,
  ],
  [
    20,
  ],
  [
    21,
  ],
  [
    22,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 6`] = `
[
  [
    22,
  ],
  [
    23,
  ],
  [
    24,
  ],
  [
    25,
  ],
  [
    26,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 7`] = `
[
  [
    26,
  ],
  [
    27,
  ],
  [
    28,
  ],
  [
    29,
  ],
  [
    30,
  ],
  [
    31,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 8`] = `
[
  [
    31,
  ],
  [
    32,
  ],
  [
    33,
  ],
  [
    34,
  ],
  [
    35,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 9`] = `
[
  [
    35,
  ],
  [
    36,
  ],
  [
    37,
  ],
  [
    38,
  ],
  [
    39,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 10`] = `
[
  [
    39,
  ],
  [
    40,
  ],
  [
    41,
  ],
  [
    42,
  ],
  [
    43,
  ],
  [
    44,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 11`] = `
[
  [
    44,
  ],
  [
    45,
  ],
  [
    46,
  ],
  [
    47,
  ],
  [
    48,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 12`] = `
[
  [
    48,
  ],
  [
    49,
  ],
  [
    50,
  ],
  [
    51,
  ],
  [
    52,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 13`] = `
[
  [
    52,
    1,
  ],
  [
    1,
    2,
  ],
  [
    2,
    3,
  ],
  [
    3,
    4,
  ],
  [
    4,
    5,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 14`] = `
[
  [
    5,
    6,
  ],
  [
    6,
    7,
  ],
  [
    7,
    8,
  ],
  [
    8,
    9,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 15`] = `
[
  [
    9,
    10,
  ],
  [
    10,
    11,
  ],
  [
    11,
    12,
  ],
  [
    12,
    13,
  ],
  [
    13,
    14,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 16`] = `
[
  [
    13,
    14,
  ],
  [
    14,
    15,
  ],
  [
    15,
    16,
  ],
  [
    16,
    17,
  ],
  [
    17,
    18,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 17`] = `
[
  [
    18,
    19,
  ],
  [
    19,
    20,
  ],
  [
    20,
    21,
  ],
  [
    21,
    22,
  ],
  [
    22,
    23,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 18`] = `
[
  [
    22,
    23,
  ],
  [
    23,
    24,
  ],
  [
    24,
    25,
  ],
  [
    25,
    26,
  ],
  [
    26,
    27,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 19`] = `
[
  [
    26,
    27,
  ],
  [
    27,
    28,
  ],
  [
    28,
    29,
  ],
  [
    29,
    30,
  ],
  [
    30,
    31,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 20`] = `
[
  [
    31,
    32,
  ],
  [
    32,
    33,
  ],
  [
    33,
    34,
  ],
  [
    34,
    35,
  ],
  [
    35,
    36,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 21`] = `
[
  [
    35,
    36,
  ],
  [
    36,
    37,
  ],
  [
    37,
    38,
  ],
  [
    38,
    39,
  ],
  [
    39,
    40,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 22`] = `
[
  [
    39,
    40,
  ],
  [
    40,
    41,
  ],
  [
    41,
    42,
  ],
  [
    42,
    43,
  ],
  [
    43,
    44,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 23`] = `
[
  [
    44,
    45,
  ],
  [
    45,
    46,
  ],
  [
    46,
    47,
  ],
  [
    47,
    48,
  ],
  [
    48,
    49,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 24`] = `
[
  [
    48,
    49,
  ],
  [
    49,
    50,
  ],
  [
    50,
    51,
  ],
  [
    51,
    52,
  ],
  [
    52,
    1,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 25`] = `
[
  [
    52,
  ],
  [
    1,
  ],
  [
    2,
  ],
  [
    3,
  ],
  [
    4,
  ],
  [
    5,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 26`] = `
[
  [
    5,
  ],
  [
    6,
  ],
  [
    7,
  ],
  [
    8,
  ],
  [
    9,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 27`] = `
[
  [
    9,
  ],
  [
    10,
  ],
  [
    11,
  ],
  [
    12,
  ],
  [
    13,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 28`] = `
[
  [
    13,
  ],
  [
    14,
  ],
  [
    15,
  ],
  [
    16,
  ],
  [
    17,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 29`] = `
[
  [
    18,
  ],
  [
    19,
  ],
  [
    20,
  ],
  [
    21,
  ],
  [
    22,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 30`] = `
[
  [
    22,
  ],
  [
    23,
  ],
  [
    24,
  ],
  [
    25,
  ],
  [
    26,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 31`] = `
[
  [
    26,
  ],
  [
    27,
  ],
  [
    28,
  ],
  [
    29,
  ],
  [
    30,
  ],
  [
    31,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 32`] = `
[
  [
    31,
  ],
  [
    32,
  ],
  [
    33,
  ],
  [
    34,
  ],
  [
    35,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 33`] = `
[
  [
    35,
  ],
  [
    36,
  ],
  [
    37,
  ],
  [
    38,
  ],
  [
    39,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 34`] = `
[
  [
    39,
  ],
  [
    40,
  ],
  [
    41,
  ],
  [
    42,
  ],
  [
    43,
  ],
  [
    44,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 35`] = `
[
  [
    44,
  ],
  [
    45,
  ],
  [
    46,
  ],
  [
    47,
  ],
  [
    48,
  ],
]
`;

exports[`generateWeeksBasedOnOverlap 36`] = `
[
  [
    48,
  ],
  [
    49,
  ],
  [
    50,
  ],
  [
    51,
  ],
  [
    52,
  ],
]
`;
